File: node-js/src/core/ExactCoverCounter.js
class Node {
  constructor() {
    this.left = this;
    this.right = this;
    this.up = this;
    this.down = this;
    this.column = null;
  }
}

class Column extends Node {
  constructor() {
    super();
    this.column = this;
    this.size = 0;
  }
}

class ExactCoverCounter {
  constructor(columnCount, rows, solutionLimit = Infinity) {
    this.root = new Column();
    this.columns = new Array(columnCount);
    this.solutions = 0;
    this.triedRows = 0;
    this.solutionLimit = solutionLimit;
    this.selectedRows = new Array(columnCount);
    this.solutionRows = [];

    let previous = this.root;
    for (let index = 0; index < columnCount; index++) {
      const column = new Column();
      this.columns[index] = column;
      this._linkHorizontally(previous, column);
      previous = column;
    }
    this._linkHorizontally(previous, this.root);

    for (let rowIndex = 0; rowIndex < rows.length; rowIndex++) {
      this._addRow(rows[rowIndex], rowIndex);
    }
  }

  static count(columnCount, rows) {
    const counter = new ExactCoverCounter(columnCount, rows);
    counter._search(0);
    return { solutions: counter.solutions, triedRows: counter.triedRows };
  }

  static findOne(columnCount, rows) {
    const counter = new ExactCoverCounter(columnCount, rows, 1);
    counter._search(0);
    return {
      solutions: counter.solutions,
      triedRows: counter.triedRows,
      selectedRows: counter.solutionRows
    };
  }

  _addRow(columnIndexes, rowIndex) {
    let first = null;
    let previous = null;

    for (const columnIndex of columnIndexes) {
      const column = this.columns[columnIndex];
      const node = new Node();
      node.column = column;
      node.rowIndex = rowIndex;

      node.down = column;
      node.up = column.up;
      column.up.down = node;
      column.up = node;
      column.size++;

      if (first === null) {
        first = node;
      } else {
        this._linkHorizontally(previous, node);
      }
      previous = node;
    }
    this._linkHorizontally(previous, first);
  }

  _search(depth) {
    if (this.root.right === this.root) {
      this.solutions++;
      this.solutionRows = this.selectedRows.slice(0, depth);
      return this.solutions >= this.solutionLimit;
    }

    const column = this._smallestColumn();
    if (column.size === 0) return false;

    this._cover(column);
    for (let row = column.down; row !== column; row = row.down) {
      this.triedRows++;
      this.selectedRows[depth] = row.rowIndex;
      for (let node = row.right; node !== row; node = node.right) {
        this._cover(node.column);
      }
      const limitReached = this._search(depth + 1);
      for (let node = row.left; node !== row; node = node.left) {
        this._uncover(node.column);
      }
      if (limitReached) {
        this._uncover(column);
        return true;
      }
    }
    this._uncover(column);
    return false;
  }

  _smallestColumn() {
    let smallest = null;
    for (let node = this.root.right; node !== this.root; node = node.right) {
      if (smallest === null || node.size < smallest.size) {
        smallest = node;
      }
    }
    return smallest;
  }

  _cover(column) {
    column.right.left = column.left;
    column.left.right = column.right;
    for (let row = column.down; row !== column; row = row.down) {
      for (let node = row.right; node !== row; node = node.right) {
        node.down.up = node.up;
        node.up.down = node.down;
        node.column.size--;
      }
    }
  }

  _uncover(column) {
    for (let row = column.up; row !== column; row = row.up) {
      for (let node = row.left; node !== row; node = node.left) {
        node.column.size++;
        node.down.up = node;
        node.up.down = node;
      }
    }
    column.right.left = column;
    column.left.right = column;
  }

  _linkHorizontally(left, right) {
    left.right = right;
    right.left = left;
  }
}

module.exports = ExactCoverCounter;

File: kotlin/src/main/kotlin/edu/generalpuzzle/core/ExactCoverCounter.kt
package edu.generalpuzzle.core

internal data class ExactCoverResult(
    val solutions: Int,
    val triedRows: Int,
    val selectedRows: List<Int> = emptyList()
)

/**
 * Knuth's Algorithm X with dancing links.
 *
 * Puzzle placements are rows and every board cell and puzzle piece is a column.
 * A complete exact cover therefore fills every cell while using every piece once.
 */
internal class ExactCoverCounter private constructor(
    columnCount: Int,
    rows: List<IntArray>,
    private val solutionLimit: Int
) {
    private open class Node {
        var left: Node = this
        var right: Node = this
        var up: Node = this
        var down: Node = this
        lateinit var column: Column
        var rowIndex = -1
    }

    private class Column : Node() {
        var size = 0

        init {
            column = this
        }
    }

    private val root = Column()
    private val columns = Array(columnCount) { Column() }
    private val selectedRows = IntArray(columnCount)
    private var solutionRows = emptyList<Int>()
    private var solutions = 0
    private var triedRows = 0

    init {
        var previous: Node = root
        for (column in columns) {
            linkHorizontally(previous, column)
            previous = column
        }
        linkHorizontally(previous, root)

        rows.forEachIndexed(::addRow)
    }

    private fun addRow(rowIndex: Int, columnIndexes: IntArray) {
        var first: Node? = null
        var previous: Node? = null

        for (columnIndex in columnIndexes) {
            val column = columns[columnIndex]
            val node = Node().apply {
                this.column = column
                this.rowIndex = rowIndex
                down = column
                up = column.up
            }
            column.up.down = node
            column.up = node
            column.size++

            if (first == null) {
                first = node
            } else {
                linkHorizontally(previous!!, node)
            }
            previous = node
        }

        if (first != null) {
            linkHorizontally(previous!!, first)
        }
    }

    private fun search(depth: Int): Boolean {
        if (Thread.currentThread().isInterrupted) {
            return true
        }
        if (root.right === root) {
            solutions++
            solutionRows = selectedRows.take(depth)
            return solutions >= solutionLimit
        }

        val column = smallestColumn()
        if (column.size == 0) {
            return false
        }

        cover(column)
        var row = column.down
        while (row !== column) {
            triedRows++
            selectedRows[depth] = row.rowIndex

            var node = row.right
            while (node !== row) {
                cover(node.column)
                node = node.right
            }

            val limitReached = search(depth + 1)

            node = row.left
            while (node !== row) {
                uncover(node.column)
                node = node.left
            }
            if (limitReached) {
                uncover(column)
                return true
            }
            row = row.down
        }
        uncover(column)
        return false
    }

    private fun smallestColumn(): Column {
        var smallest: Column? = null
        var node = root.right
        while (node !== root) {
            val column = node as Column
            if (smallest == null || column.size < smallest.size) {
                smallest = column
            }
            node = node.right
        }
        return smallest!!
    }

    private fun cover(column: Column) {
        column.right.left = column.left
        column.left.right = column.right
        var row = column.down
        while (row !== column) {
            var node = row.right
            while (node !== row) {
                node.down.up = node.up
                node.up.down = node.down
                node.column.size--
                node = node.right
            }
            row = row.down
        }
    }

    private fun uncover(column: Column) {
        var row = column.up
        while (row !== column) {
            var node = row.left
            while (node !== row) {
                node.column.size++
                node.down.up = node
                node.up.down = node
                node = node.left
            }
            row = row.up
        }
        column.right.left = column
        column.left.right = column
    }

    private fun linkHorizontally(left: Node, right: Node) {
        left.right = right
        right.left = left
    }

    companion object {
        fun count(columnCount: Int, rows: List<IntArray>): ExactCoverResult {
            return run(columnCount, rows, Int.MAX_VALUE)
        }

        fun findOne(columnCount: Int, rows: List<IntArray>): ExactCoverResult {
            return run(columnCount, rows, 1)
        }

        private fun run(columnCount: Int, rows: List<IntArray>, limit: Int): ExactCoverResult {
            val counter = ExactCoverCounter(columnCount, rows, limit)
            counter.search(0)
            return ExactCoverResult(counter.solutions, counter.triedRows, counter.solutionRows)
        }
    }
}
